extensions/amp-video/0.1/amp-video-base.js

```diff
@@ -0,0 +1,747 @@
+/**
+  * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+  *
+  * Licensed under the Apache License, Version 2.0 (the "License");
+  * you may not use this file except in compliance with the License.
+  * You may obtain a copy of the License at
+  *
+  *      http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS-IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+import {EMPTY_METADATA} from '../../../src/mediasession-helper';
+import {Services} from '../../../src/services';
+import {VideoEvents} from '../../../src/video-interface';
+import {VisibilityState} from '../../../src/visibility-state';
+import {
+  childElementByTag,
+  childElementsByTag,
+  elementByTag,
+  fullscreenEnter,
+  fullscreenExit,
+  insertAfterOrAtStart,
+  isFullscreenElement,
+  removeElement,
+} from '../../../src/dom';
+import {descendsFromStory} from '../../../src/utils/story';
+import {dev, devAssert} from '../../../src/log';
+import {getMode} from '../../../src/mode';
+import {htmlFor} from '../../../src/static-template';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {listen} from '../../../src/event-helper';
+import {mutedOrUnmutedEvent} from '../../../src/iframe-video';
+import {
+  propagateObjectFitStyles,
+  setImportantStyles,
+  setInitialDisplay,
+  setStyles,
+} from '../../../src/style';
+import {toArray} from '../../../src/types';
+
+/** @protected {!Array<string>} */
+const ATTRS_TO_PROPAGATE_ON_BUILD = [
+  'aria-describedby',
+  'aria-label',
+  'aria-labelledby',
+  'controls',
+  'crossorigin',
+  'disableremoteplayback',
+  'poster',
+  'controlsList',
+];
+
+/**
+ * @note Do not propagate `autoplay`. Autoplay behavior is managed by
+ *       video manager since amp-video implements the VideoInterface.
+ * @protected {!Array<string>}
+ */
+const ATTRS_TO_PROPAGATE_ON_LAYOUT = ['loop', 'preload'];
+
+/** @protected {!Array<string>} */
+const ATTRS_TO_PROPAGATE =
+    ATTRS_TO_PROPAGATE_ON_BUILD.concat(ATTRS_TO_PROPAGATE_ON_LAYOUT);
+
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+export class AmpVideoBase extends AMP.BaseElement {
```

Comment:
Please **do not** extend components, *specially* across extensions.

Component hierarchies are inherently complex (e.g. see Closure UI) and not a good precedent to set. This also tightly couples `amp-video` with the shaka player, which is a completely independent extension.

If you need similar behavior, please refactor into common utilities that can be composed.